lightly/models/modules/masked_vision_transformer_timm.py

```diff
@@ -0,0 +1,140 @@
+        # normalization layer
+        input = self.vit.norm_pre(input)
+        # apply Transformer blocks
+        input = self.vit.blocks(input)
+        # normalize
+        out: torch.Tensor = self.vit.norm(input)
+        return out
+
+    def patch_embed(self, images: Tensor) -> Tensor:
+        return self.vit.patch_embed(images)
+
+    def add_prefix_tokens(self, x: Tensor) -> Tensor:
+        tokens = torch.Tensor([])
+        tokens = tokens.to(self.device)
+        if self.vit.cls_token is not None:
+            tokens = torch.cat((tokens, self.vit.cls_token.expand(x.shape[0], -1, -1)), dim=1)
+        if self.vit.reg_token is not None:
+            tokens = torch.cat((tokens, self.vit.reg_token.expand(x.shape[0], -1, -1)), dim=1)
+        return torch.cat((tokens, x), dim=1)
+
+
+    def add_pos_embed(self, x: Tensor) -> Tensor:
+        if self.vit.dynamic_img_size:
+            B, H, W, C = x.shape
+            pos_embed = resample_abs_pos_embed(
+                self.vit.pos_embed,
+                (H, W),
+                num_prefix_tokens=0 if self.vit.no_embed_class else self.vit.num_prefix_tokens,
+            )
+            x = x.view(B, -1, C)
+        else:
+            pos_embed = self.vit.pos_embed
+        # TODO: give option to either have positional encoding for the prefix tokens or not using the self.vit.no_embed_class. 
+        # Here it is assumed that the prefix tokens always have positional encodings.
+        x = x + pos_embed
+        return self.vit.pos_drop(x)
+    
+    def _initialize_weights(self) -> None:
+        # Initialize the patch embedding layer like a linear layer instead of conv
+        # layer.
+        w = self.patch_embed.proj.weight.data
+        torch.nn.init.xavier_uniform_(w.view([w.shape[0], -1]))
+
+        # Initialize the class token.
+        torch.nn.init.normal_(self.cls_token, std=0.02)
+
+        # initialize nn.Linear and nn.LayerNorm
+        self.apply(_init_weights)
+
+        _initialize_2d_sine_cosine_positional_embedding(self.vit.pos_embed)
+        
+        
+def _initialize_2d_sine_cosine_positional_embedding(pos_embedding: Parameter) -> None:
+    _, seq_length, hidden_dim = pos_embedding.shape
+    grid_size = int((seq_length - 1) ** 0.5)
+    sine_cosine_embedding = utils.get_2d_sine_cosine_positional_embedding(
+        embed_dim=hidden_dim,
+        grid_size=grid_size,
+        cls_token=True,
+    )
+    pos_embedding.data.copy_(
+        torch.from_numpy(sine_cosine_embedding).float().unsqueeze(0)
+    )
+    # Freeze positional embedding.
+    pos_embedding.requires_grad = False
+
+
+def _init_weights(module: Module) -> None:
+    if isinstance(module, Linear):
+        nn.init.xavier_uniform_(module.weight)
+        if isinstance(module, Linear) and module.bias is not None:
+            nn.init.constant_(module.bias, 0)
+    elif isinstance(module, LayerNorm):
+        nn.init.constant_(module.bias, 0)
+        nn.init.constant_(module.weight, 1.0)
```

Comment:
Weight initialization should already be handled by the `VisionTransformer` superclass. I think we should move the MAE specific initialization outside the class. For example we can make a function `init_weights_mae` that does all this. And I think we can move `_initialize_2d_sine_cosine_positional_embedding` to `models/utils.py`  as we already have the other positional embedding functions there.

Reply:
I don't think that the weight initialization can be handled by the `VisionTransformer` class because depending on whether we use torchvision or TIMM for example the methods would be different. I think I will keep the initialization as it is for now, because this is what was being done essentially so far (when the `Backbone` of the `masked_autoencoder.py` was used for other architectures). It will also make more straightforward the comparison of the examples with this new backbone and with the one that we had before. WDYT?

Reply:
We can leave it for now but eventually we should move it outside the class (not outside the file though!). The reason is that different methods use different initialization schemes. Having a single `_init_weights` method will not cover the different cases.